### PR TITLE
Explore: Create DataSourceWithSupplementaryQueriesSupport interface to support log volume and samples

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What is this feature?**

[Logs sample](https://www.figma.com/file/YeTF07V4TUNydMJMD1DYBd/Query-builder-improvements?node-id=4149%3A27563&t=NBuDUIeRcDLYcUFk-0) is a new feature in Explore. It is intended to work in the same way as Logs volume, but in reverse. If you run metric query, this will allow you to see log samples.

In this PR, I am focusing solely on adding logic to create logs sample request, process and store data. The UI part will come in the next PR. 

I have also implemented log samples logic for Loki.

**Why do we need this feature?**

This feature is for users who use Explore to interact with the data sources that support both - logs and metrics. It gives them a way to see log samples for when they are running metrics query. 

More on this in https://www.figma.com/file/YeTF07V4TUNydMJMD1DYBd/Query-builder-improvements?node-id=4149%3A27563&t=NBuDUIeRcDLYcUFk-0

**Which issue(s) does this PR fix?**:

Part of https://github.com/grafana/grafana/issues/52088

**Special notes for your reviewer**:

As this PR is still quite big, I recommend to review it commit-by-commit, I tried to split it into smaller and easier to review commits. If it is still too much, let me know and I can create more smaller PRs. 

There is also 1 tiny breaking change. Instead of storing localStorage settings keys individually, I have created a utility function that creates it from the supp query type. However, the naming for logs volume was `grafana.explore.logs.enableVolumeHistogram'` which I wasn't able to keep. This means that user who toggled off logs volume will see it and will have ti disable it again. Let me know if you agree that it is okay, or if I should keep it and use `getSupplementaryQuerySettingKey` just for the new query. 

```
const getSupplementaryQuerySettingKey = (type: SupplementaryQueryType) => `grafana.explore.logs.enable${type}`;
```

### CODE REVIEW

File: public/app/core/logsModel.ts
Comment (matyax) [inline public/app/core/logsModel.ts:759]: Do we really need to use observables here? I can't find the use case and the produced code is very obscure.

File: public/app/features/explore/state/utils.ts
Comment (matyax) [inline public/app/features/explore/state/utils.ts:80]: Why can't we use an actual `TimeRange` instance here? Because this is not a test. Same for `AbsoluteTimeRange`.

**Review by matyax — COMMENTED** [review]
Did a first pass and left some comments.

File: public/app/features/explore/state/query.ts
Comment (ifrost) [inline public/app/features/explore/state/query.ts:620]: Is it possible to make it more generic so we don't have `if` for each type, but rather iterate over all of them?

Comment (ivanahuckova) [reply public/app/features/explore/state/query.ts]: Good idea, I refactored it in here https://github.com/grafana/grafana/pull/61298/commits/e9a6a1d3c66b4d5856ea833345ace1afea4e8706. Let me know what do you think!

File: public/app/features/explore/state/utils.ts
Comment (ivanahuckova) [reply public/app/features/explore/state/utils.ts]: I added this fix thinking it is a test file. I removed this changes in https://github.com/grafana/grafana/pull/61298/commits/ff905e3773a715b1548971d203d163389cd1d873

File: public/app/features/explore/state/query.ts
Comment (ivanahuckova) [reply public/app/features/explore/state/query.ts]: Good idea! I have refactored it in https://github.com/grafana/grafana/pull/61298/commits/e9a6a1d3c66b4d5856ea833345ace1afea4e8706. (I removed redundant Object.values when looping over supplementaryQueriesList in https://github.com/grafana/grafana/pull/61298/commits/41abe6411b9b44a77526ab713d04be7d492785b7. I forgot to fix it in first commit.) And then I fixed additional bug in logic in https://github.com/grafana/grafana/pull/61298/commits/b31a65fabed8c3412ab82b647a1a7584425651f8.

Let me know what do you think

File: public/app/features/explore/state/query.test.ts
Comment (ifrost) [inline public/app/features/explore/state/query.test.ts:572]: That's a lot of copy-pasta 🍝  Not sure if we need it since you made the logic more generic 🤔  Or maybe we could have one mock that implements both providers and merge all tests cases?

Comment (ifrost) [reply public/app/features/explore/state/query.test.ts:572]: Also if we had unified interface mentioned earlier (DataSourceWithSupplementaryQueriesSupport) we could make the test simpler.

**Review by ifrost — COMMENTED** [review]
Nice work! Seems to work as expected -> tested by checking what queries are run in the network tab.

Regarding local storage key: would it make sense to add migration for the key? And remove it in G10?

Also: Could running queries in background potentially cause some extra cost for the user? If yes, I'd suggest putting this behind a feature toggle until docs/UI are updated.

File: public/app/features/explore/state/query.test.ts
Comment (Elfo404) [inline public/app/features/explore/state/query.test.ts:617]: same as before for all the added tests here, `dispatch` is not async, there's no need to `await` it or mark the test as async

**Review by Elfo404 — COMMENTED** [review]
seems to work as expected, left a couple of (not so) nits, otherwise lgtm. would wait for @ifrost to approve as he has more context

**Comment by ivanahuckova** [comment]
> Regarding local storage key: would it make sense to add migration for the key? And remove it in G10?

@ifrost good idea, added in  https://github.com/grafana/grafana/pull/61298/commits/896edaf7a4414bf5d5265cfe66adbb12a8e369bd

>Also: Could running queries in background potentially cause some extra cost for the user? If yes, I'd suggest putting this behind a feature toggle until docs/UI are updated.

No, this definitely isn't as problematic as when we were solving logs volume.

File: public/app/features/explore/utils/supplementaryQueries.ts
Comment (ifrost) [inline public/app/features/explore/utils/supplementaryQueries.ts:63]:
```suggestion
      // TODO: Remove this in 10.0 (#61626)
```

File: packages/grafana-data/src/types/index.ts
Comment (academo) [inline packages/grafana-data/src/types/index.ts:10]: @ivanahuckova removing this file from grafana-data breaks our plugins API. If you are planning on removing this and replacing them for others, please follow the procedure of deprecating them in this PR and only removing them in the next major version.

Comment (ivanahuckova) [reply packages/grafana-data/src/types/index.ts:10]: Hello @academo! Is this needed as `DataSourceWithLogsVolumeSupport` API had `@internal` tag and we suspected that it might change?  We replaced this with more generic `DataSourceWithSupplementaryQueriesSupport` and implemented change for all internal plugins that were using `DataSourceWithLogsVolumeSupport`.

**Review by ifrost — APPROVED** [review]

File: packages/grafana-data/src/types/logs.ts
Comment (ivanahuckova) [reply packages/grafana-data/src/types/logs.ts]: ✨

File: packages/grafana-data/src/types/index.ts
Comment (academo) [reply packages/grafana-data/src/types/index.ts:10]: Hi @ivanahuckova I checked the information we have on API usage and we could not find any external plugins using it. so go ahead with merging this PR

Comment (tolzhabayev) [reply packages/grafana-data/src/types/index.ts:10]: @ivanahuckova 


- https://github.com/grafana/newrelic-datasource/blob/main/src/datasource/Datasource.ts#L99 
- https://github.com/TencentCloud/tencentcloud-monitor-grafana-app/blob/1b1b240989b27fe111498539cbc109ef8ac97ff2/src/datasource/log-service/LogServiceDataSource.ts#L198

Comment (ivanahuckova) [reply packages/grafana-data/src/types/index.ts:10]: The [TencentCloud/](https://github.com/TencentCloud/tencentcloud-monitor-grafana-app/blob/1b1b240989b27fe111498539cbc109ef8ac97ff2/src/datasource/log-service/LogServiceDataSource.ts#L198) is not implementing an API, so there is no issue. 

However [newrelic](https://github.com/grafana/newrelic-datasource/blob/main/src/datasource/Datasource.ts#L99) is implementing and using it (fyi @yesoreyeram). 

Considering this API was marked as `@internal`, would you recommend to:
- keep this PR in this form and let NewRelic data source maintainers know about this
- keep this PR in this form, but add `breaking change` label and let NewRelic data source maintainers know about this
- Change PR to deprecate old API and introduce new one. And then in v10, break the old API
- Something else

**Comment by ivanahuckova** [comment]
@ifrost Could you please have 1 more look at https://github.com/grafana/grafana/pull/61298/commits/f5be7ad433a062d94fad18d9d08da3435382cac9 that brings back `DataSourceWithLogsVolumeSupport`, but deprecates it. Plus I've brought back logic to `query.ts` that handles logic for `logsVolume` trough `hasLogsVolumeSupport`. I did it in a way, where we will be able to just delete whole `else if block` after we remove the support in 10.

File: packages/grafana-data/src/types/index.ts
Comment (ivanahuckova) [reply packages/grafana-data/src/types/index.ts]: 👍 I brought `DataSourceWithLogsVolumeSupport` back and deprecated it: https://github.com/grafana/grafana/commit/f5be7ad433a062d94fad18d9d08da3435382cac9.

File: public/app/features/explore/state/query.ts
Comment (ifrost) [inline public/app/features/explore/state/query.ts:618]: Did you mean?

```suggestion
            // TODO: Remove after v10 (#61296)
```

Comment (ivanahuckova) [reply public/app/features/explore/state/query.ts:618]: I wanted to link this PR as reference to what should be removed.

Comment (ivanahuckova) [reply public/app/features/explore/state/query.ts:618]: But I can create issue for us - log squad and link that

Comment (ivanahuckova) [reply public/app/features/explore/state/query.ts]: Changed in https://github.com/grafana/grafana/pull/61298/commits/578d9b0f20abd68d9e698bd8a709f9d8716ca5b6 and new issue is here https://github.com/grafana/grafana/issues/61845

**Review by ifrost — APPROVED** [review]

**Pull request merged**
